### PR TITLE
Aco agent multi map movement

### PR DESCRIPTION
**Pull request by trav-d13**
Working and checked. Two bug fixes completed

### CODE REVIEW

File: src/main/java/app/model/agents/ACO/AcoAgent.java [hunk 1]
```diff
@@ -1,43 +1,228 @@
...
+            if(cardinalRays[i].rayLength() > cellSize + epsilon)
+            {
+                possibleMovements.add(angleToGridMovementLink(cardinalAngles[i]));
+            }
+        }
+        return possibleMovements;
     }
+
+    public Vector angleToGridMovementLink(int angle)
+    {
```
Comment (S010MON) [inline src/main/java/app/model/agents/ACO/AcoAgent.java:135]: This is a very nice method! Clean

File: src/main/java/app/model/agents/ACO/AcoAgent.java [hunk 2]
```diff
@@ -1,43 +1,228 @@
...
+            else if(currentAngle < targetCardinalAngle)
+            {
+                lowerBound = midPoint + 1;
+            }
+            else if(currentAngle > targetCardinalAngle)
+            {
+                upperBound = midPoint - 1;
+            }
+        }
+        System.out.println("Cardinal Point not found");
```
Comment (S010MON) [inline src/main/java/app/model/agents/ACO/AcoAgent.java:185]: Consider a runtime error instead?
```java
throw new RuntimeException("Cardinal Point not found");
```

File: src/main/java/app/model/agents/ACO/AcoAgent.java [hunk 3]
```diff
@@ -1,43 +1,228 @@
...
+    }
+
+    public int calculateMidPoint(int upperBound, int lowerBound)
+    {
+        return lowerBound + (upperBound - lowerBound)/2;
+    }
+
+    public boolean approximateAngleRange(double detectedAngle, double targetAngle)
+    {
+        if(detectedAngle < (targetAngle + epsilon) && detectedAngle > (targetAngle - epsilon))
```
Comment (S010MON) [inline src/main/java/app/model/agents/ACO/AcoAgent.java:196]: change to return value directly
```java
return detectedAngle < (targetAngle + epsilon) && detectedAngle > (targetAngle - epsilon);
```

File: src/main/java/app/model/agents/ACO/AcoAgent.java [hunk 4]
```diff
@@ -1,43 +1,228 @@
...
+    }
+
+    public Vector selectRandomEquivalentMove(ArrayList<Vector> equivalentMoves)
+    {
+        return equivalentMoves.get(randomGenerator.nextInt(equivalentMoves.size()));
+    }
+
+    public ArrayList<Double> accessAvaliableCellPheromones(ArrayList<Vector> possibleMovements)
+    {
+
```
Comment (S010MON) [inline src/main/java/app/model/agents/ACO/AcoAgent.java:108]: random space?

**Review by S010MON — APPROVED** [review]
All of your methods are public, is this required?  Make any implementation specific methods private (as much as possible) to increase encapsulation.

The code looks good and passes all the tests; I haven't looked through the tests in detail, but approved!

File: src/main/java/app/model/agents/ACO/AcoAgent.java (hunk 3 above)
Comment (S010MON) [reply src/main/java/app/model/agents/ACO/AcoAgent.java:196]: Rather than

```java
if(condition)
   return true;
else
   return false;
```

**Review by MatthewFoster02 — APPROVED** [review]
Get cardinal point not found error when running but the style and code looks good.

File: src/main/java/app/controller/linAlg/Vector.java [hunk 5]
```diff
@@ -68,7 +70,15 @@ public Vector rotate(double degrees)
     public double getAngle()
     {
         // + 90 to set the angle to north, negate angle to make it clockwise
-        return (-Math.toDegrees(Math.atan2(y,x)) + 360.0 + 90.0) % 360.0;
+        // Rounded to 1 decimal place
+        double angle = (-Math.toDegrees(Math.atan2(y,x)) + 360.0 + 90.0) % 360.0;
+        double roundedAngle = (double) Math.round(angle * 10) / 10;
+
+        if(roundedAngle == 360.0)
```
Comment (Jolijn2020) [inline src/main/java/app/controller/linAlg/Vector.java:77]: don't you need to take into account that doubles aren't precise?

for example:

if( Math.abs(roundedAngle-360.0) < 0.00001 )
```suggestion
        if(Math.abs(roundedAngle-360.0) < 0.00001 )```

File: src/main/java/app/model/Grid/Grid.java [hunk 6]
```diff
@@ -67,4 +79,9 @@ public int getColDimension()
     {
         return grid[0].length;
     }
+
+    public double getCellSize()
```
Comment (Jolijn2020) [inline src/main/java/app/model/Grid/Grid.java:83]: can use a @Getter for this

**Review by S010MON — CHANGES_REQUESTED** [review]
I'm getting a consistent error:
```java
Cardinal Point not found
Cardinal Point not found
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException: Cannot invoke "app.controller.graphicsEngine.Ray.rayLength()" because "cardinalRays[i]" is null
	at app@1.0-SNAPSHOT/app.model.agents.ACO.AcoAgent.determineAvailableMovements(AcoAgent.java:172)
	at app@1.0-SNAPSHOT/app.model.agents.ACO.AcoAgent.determineMovesAllScenarios(AcoAgent.java:88)
	at app@1.0-SNAPSHOT/app.model.agents.ACO.AcoAgent.move(AcoAgent.java:55)
	at app@1.0-SNAPSHOT/app.controller.GameEngine.tick(GameEngine.java:42)

```

It seems that the addition of throwing an exception has exposed a problem that was handled with print statements.  
> Consider a runtime error instead?
> 
> ```java
> throw new RuntimeException("Cardinal Point not found");
> ```

Up to you to advise if this is a serious error or trivial and should be returned to a `System.out.println` for Info logging

**Review by S010MON — DISMISSED** [review]

**Review by MatthewFoster02 — DISMISSED** [review]

**Comment by S010MON** [comment]
Dude ... two people reviewed this request and you just left it, now it's out of date with conflicts.  Not cool

**Pull request merged**